lib/proxy/client.go

```diff
@@ -0,0 +1,558 @@
+		c.metrics.reportTunnelError(errorExistingTunnelInvalid)
+		return nil, false, trace.NewAggregate(errs...)
+	}
+
+	// proxy does not exist in cache.
+	// get list of proxies directly from auth.
+	c.Lock()
+	defer c.Unlock()
+	proxies, err := c.config.AuthClient.GetProxies()
+	if err != nil {
+		c.metrics.reportTunnelError(errorAuthClient)
+		return nil, false, trace.NewAggregate(errs...)
+	}
+
+	errs = nil
+	for _, proxy := range proxies {
+		id := proxy.GetName()
+		if _, ok := ids[id]; !ok {
+			continue
+		}
+
+		conn, err := c.connect(id, proxy.GetPeerAddr())
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		stream, err := c.startStream(conn)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		c.conns[id] = conn
+		return stream, false, nil
+	}
+
+	c.metrics.reportTunnelError(errorProxiesUnavailable)
+	return nil, false, trace.NotFound("Error dialling all proxies: %+v", trace.NewAggregate(errs...).Error())
+}
+
+// connect dials a new connection to proxyAddr.
+func (c *Client) connect(id string, proxyPeerAddr string) (*clientConn, error) {
+	tlsConfig, err := c.config.getConfigForServer()
+	if err != nil {
+		c.metrics.reportTunnelError(errorNewTunnelDial)
+		return nil, trace.Wrap(err, "Error updating client tls config")
+	}
+
+	connCtx, cancel := context.WithCancel(c.ctx)
+	wg := new(sync.WaitGroup)
+
+	transportCreds := newProxyCredentials(credentials.NewTLS(tlsConfig))
+	conn, err := grpc.DialContext(
+		connCtx,
+		proxyPeerAddr,
+		grpc.WithTransportCredentials(transportCreds),
+		grpc.WithStatsHandler(newStatsHandler(c.metrics)),
+		grpc.WithChainStreamInterceptor(metadata.StreamClientInterceptor, utils.GRPCClientStreamErrorInterceptor, streamCounterInterceptor(wg)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                peerKeepAlive,
+			Timeout:             peerTimeout,
+			PermitWithoutStream: true,
+		}),
+	)
+	if err != nil {
+		c.metrics.reportTunnelError(errorNewTunnelDial)
+		return nil, trace.Wrap(err, "Error dialling proxy %+v", id)
+	}
+
+	return &clientConn{
+		ClientConn: conn,
+		ctx:        connCtx,
+		cancel:     cancel,
+		wg:         wg,
+		id:         id,
+		addr:       proxyPeerAddr,
+	}, nil
+}
+
+// startStream opens a new stream to the provided connection.
+func (c *Client) startStream(conn *clientConn) (clientapi.ProxyService_DialNodeClient, error) {
+	client := clientapi.NewProxyServiceClient(conn.ClientConn)
+
+	stream, err := client.DialNode(conn.ctx)
+	if err != nil {
+		c.metrics.reportTunnelError(errorStreamStart)
+		return nil, trace.Wrap(err, "Error opening stream to proxy %+v", conn.id)
+	}
+
+	go func() {
+		<-conn.ctx.Done()
+		if err := stream.CloseSend(); err != nil {
+			c.config.Log.Debugf("error closing stream: %+v", err)
+		}
+	}()
+
+	return stream, nil
+}
+
+// testConnection opens a new stream to the provided connection and
+// immediately closes it.
+func (c *Client) testConnection(conn *clientConn) error {
```

Comment:
You're right it is not necessary, but it's mostly useful for reporting:
when dialling a new proxy, the only ways for us to know and report connection issues are:
- add the `WaitForReady` flag in the dial options. That waits for the duration set for timeouts before reporting unreachable servers, and that can be too long when dialing a proxy through updates or individually triggered when dialing a proxy that's not in the cache.
- without the `WaitForReady` flag. We'd have to wait for the first time a connection is being used to report it hasn't been established properly or have a routine monitoring connections.

Well I chose to initiate that test manually to get a sense if the connection has been established or not 😄. We can go with the monitoring routine if you feel that's better or even do the testing in a separate routine just for reporting purposes...

Reply:
I think the goroutine to monitor things would be better. It could just loop at some interval checking the state of each connection.

The main benefit I see is that if we discover several new proxies at once they can be added to the map faster rather than waiting while each connection is established and tested.

Also we would probably want a goroutine to monitor them regardless of whether we test them or not. Otherwise we would only update the metrics when a new set of proxies comes in or we attempt to dial over the connection.

Reply:
> Also we would probably want a goroutine to monitor them regardless of whether we test them or not. Otherwise we would only update the metrics when a new set of proxies comes in or we attempt to dial over the connection.

Now that I think about this more, I kind of disagree with it a bit. 
- It is possible we miss reporting connections being down if the ticker delay isn't small enough. In that case what would you think would be an acceptable interval?
- We're also probably not interested in these kinds of metrics, at least I am not 😄. What I care most about is mostly of the following order; were we able to dial a proxy we've been told is reachable? 

But on the other side I don't have a really strong opinion on either approaches. I'll rework the code a bit and see how it plays with the `loadBalancingPolicy` config